### ISSUE TITLE
Building vim with python3.10 (current cpython HEAD) does not configure correctly

### ISSUE THREAD

**Issue by xr1s** [vim/vim#6183]
The corresponding lines of config.log from configuring.
```
configure:6875: checking if compile and link flags for Python 3 are sane
configure:6892: gcc -o conftest -g -O2 -I/home/xris/.local/include/python3.1 -pthread  -L/home/xris/.local/lib -L/home/xris/.local/lib64 -lncursesw -ltinfow -L/usr/local/lib conftest.c    -L/home/xris/.local/lib/python3.10/config-3.10-x86_64-linux-gnu -lpython3.1 -lcrypt -lpthread -ldl -lutil -lm -lm >&5
/home/xris/.local/bin/../lib/gcc/x86_64-pc-linux-gnu/11.0.0/../../../../x86_64-pc-linux-gnu/bin/ld: cannot find -lpython3.1
```

As shown in config.log, the configure incorrectly treated python3.10 as python3.1 and tried to link with `-lpython3.1`.

I checked src/configure.ac and found in line 1427, the configure file tried to get python version from `import sys; sys.version[:3]`, which will definitely fail when minor version has two digits. My suggestion is to use `sys.version_info.major` and `sys.version_info.minor`, both are available since python 2.0.

**Environment:**
 - VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jun  2 2020 10:12:44)
 - OS: Debian GNU/Linux 9 (stretch)
 - Python: Python 3.10.0a0 (heads/master-dirty:8a3d2af997, Jun  2 2020, 09:56:18)

**Comment by xr1s** [comment]
```

**Comment by xr1s** [comment]
A possible fix to configure.

Generate configure file with `autoconf (GNU Autoconf) 2.69`

And I could not find symbols `PyRun_SimpleString` and `PyRun_String` in libpython3.10.a (and source code, I diff'ed 3.10 and 3.9 and found the definition of those two functions were removed, only macros left), so patched src/if_python3.c at the same time.

Works well with YouCompleteMe.

```
diff --git a/src/auto/configure b/src/auto/configure
index 9c9a719d8..c43f23af0 100755
--- a/src/auto/configure
+++ b/src/auto/configure
@@ -6676,20 +6676,33 @@ done

         { $as_echo "$as_me:${as_lineno-$LINENO}: checking Python version" >&5
 $as_echo_n "checking Python version... " >&6; }
-if ${vi_cv_var_python3_version+:} false; then :
+if ${vi_cv_var_python3_version_major+:} false; then :
   $as_echo_n "(cached) " >&6
 else
-  vi_cv_var_python3_version=`
-          ${vi_cv_path_python3} -c 'import sys; print(sys.version[:3])'`
+  vi_cv_var_python3_version_major=`
+          ${vi_cv_path_python3} -c 'import sys; print(sys.version_info.major)'`

 fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $vi_cv_var_python3_version" >&5
-$as_echo "$vi_cv_var_python3_version" >&6; }
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $vi_cv_var_python3_version_major" >&5
+$as_echo "$vi_cv_var_python3_version_major" >&6; }
+    { $as_echo "$as_me:${as_lineno-$LINENO}: checking Python version" >&5
+$as_echo_n "checking Python version... " >&6; }
+if ${vi_cv_var_python3_version_minor+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+  vi_cv_var_python3_version_minor=`
+          ${vi_cv_path_python3} -c 'import sys; print(sys.version_info.minor)'`
+
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $vi_cv_var_python3_version_minor" >&5
+$as_echo "$vi_cv_var_python3_version_minor" >&6; }
+    vi_cv_var_python3_version=${vi_cv_var_python3_version_major}.${vi_cv_var_python3_version_minor}

         { $as_echo "$as_me:${as_lineno-$LINENO}: checking Python is 3.0 or better" >&5
 $as_echo_n "checking Python is 3.0 or better... " >&6; }
     if ${vi_cv_path_python3} -c \
-      "import sys; sys.exit(${vi_cv_var_python3_version} < 3.0)"
+      "import sys; sys.exit((${vi_cv_var_python3_version_major},
+                            ${vi_cv_var_python3_version_minor}) < (3, 0))"
     then
       { $as_echo "$as_me:${as_lineno-$LINENO}: result: yep" >&5
 $as_echo "yep" >&6; }
@@ -6702,7 +6715,8 @@ else

        vi_cv_var_python3_abiflags=
        if ${vi_cv_path_python3} -c \
-           "import sys; sys.exit(${vi_cv_var_python3_version} < 3.2)"
+           "import sys; sys.exit((${vi_cv_var_python3_version_major},
+                                 ${vi_cv_var_python3_version_minor}) < (3, 2))"
        then
          vi_cv_var_python3_abiflags=`${vi_cv_path_python3} -c \
            "import sys; print(sys.abiflags)"`
diff --git a/src/configure.ac b/src/configure.ac
index 19ce4c7a5..974045d07 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -1434,15 +1434,21 @@ if test "$enable_python3interp" = "yes" -o "$enable_python3interp" = "dynamic";
   if test "X$vi_cv_path_python3" != "X"; then

     dnl -- get its version number
-    AC_CACHE_CHECK(Python version,vi_cv_var_python3_version,
-    [[vi_cv_var_python3_version=`
-          ${vi_cv_path_python3} -c 'import sys; print(sys.version[:3])'`
+    AC_CACHE_CHECK(Python version,vi_cv_var_python3_version_major,
+    [[vi_cv_var_python3_version_major=`
+          ${vi_cv_path_python3} -c 'import sys; print(sys.version_info.major)'`
     ]])
+    AC_CACHE_CHECK(Python version,vi_cv_var_python3_version_minor,
+    [[vi_cv_var_python3_version_minor=`
+          ${vi_cv_path_python3} -c 'import sys; print(sys.version_info.minor)'`
+    ]])
+    vi_cv_var_python3_version=${vi_cv_var_python3_version_major}.${vi_cv_var_python3_version_minor}

     dnl -- it must be at least version 3
     AC_MSG_CHECKING(Python is 3.0 or better)
     if ${vi_cv_path_python3} -c \
-      "import sys; sys.exit(${vi_cv_var_python3_version} < 3.0)"
+      "import sys; sys.exit((${vi_cv_var_python3_version_major},
+                            ${vi_cv_var_python3_version_minor}) < (3, 0))"
     then
       AC_MSG_RESULT(yep)

@@ -1451,7 +1457,8 @@ if test "$enable_python3interp" = "yes" -o "$enable_python3interp" = "dynamic";
       [
        vi_cv_var_python3_abiflags=
        if ${vi_cv_path_python3} -c \
-           "import sys; sys.exit(${vi_cv_var_python3_version} < 3.2)"
+           "import sys; sys.exit((${vi_cv_var_python3_version_major},
+                                 ${vi_cv_var_python3_version_minor}) < (3, 2))"
        then
          vi_cv_var_python3_abiflags=`${vi_cv_path_python3} -c \
            "import sys; print(sys.abiflags)"`
diff --git a/src/if_python3.c b/src/if_python3.c
index 843fa079f..e37f332b3 100644
--- a/src/if_python3.c
+++ b/src/if_python3.c
@@ -188,9 +188,9 @@ typedef PySliceObject PySliceObject_T;
 # define PyObject_IsTrue py3_PyObject_IsTrue
 # define PyModule_GetDict py3_PyModule_GetDict
 #undef PyRun_SimpleString
-# define PyRun_SimpleString py3_PyRun_SimpleString
+# define PyRun_SimpleString(command) py3_PyRun_SimpleStringFlags(command, NULL)
 #undef PyRun_String
-# define PyRun_String py3_PyRun_String
+# define PyRun_String(str, start, globals, locals) py3_PyRun_StringFlags(str, start, globals, locals, NULL)
 # define PyObject_GetAttrString py3_PyObject_GetAttrString
 # define PyObject_HasAttrString py3_PyObject_HasAttrString
 # define PyObject_SetAttrString py3_PyObject_SetAttrString
@@ -334,8 +334,8 @@ static void (*py3_Py_Finalize)(void);
 static void (*py3_PyErr_SetString)(PyObject *, const char *);
 static void (*py3_PyErr_SetObject)(PyObject *, PyObject *);
 static int (*py3_PyErr_ExceptionMatches)(PyObject *);
-static int (*py3_PyRun_SimpleString)(char *);
-static PyObject* (*py3_PyRun_String)(char *, int, PyObject *, PyObject *);
+static int (*py3_PyRun_SimpleStringFlags)(char *, PyCompilerFlags *);
+static PyObject* (*py3_PyRun_StringFlags)(char *, int, PyObject *, PyObject *, PyCompilerFlags *);
 static PyObject* (*py3_PyObject_GetAttrString)(PyObject *, const char *);
 static int (*py3_PyObject_HasAttrString)(PyObject *, const char *);
 static int (*py3_PyObject_SetAttrString)(PyObject *, const char *, PyObject *);
@@ -514,8 +514,8 @@ static struct
     {"PyErr_SetString", (PYTHON_PROC*)&py3_PyErr_SetString},
     {"PyErr_SetObject", (PYTHON_PROC*)&py3_PyErr_SetObject},
     {"PyErr_ExceptionMatches", (PYTHON_PROC*)&py3_PyErr_ExceptionMatches},
-    {"PyRun_SimpleString", (PYTHON_PROC*)&py3_PyRun_SimpleString},
-    {"PyRun_String", (PYTHON_PROC*)&py3_PyRun_String},
+    {"PyRun_SimpleStringFlags", (PYTHON_PROC*)&py3_PyRun_SimpleStringFlags},
+    {"PyRun_StringFlags", (PYTHON_PROC*)&py3_PyRun_StringFlags},
     {"PyObject_GetAttrString", (PYTHON_PROC*)&py3_PyObject_GetAttrString},
     {"PyObject_HasAttrString", (PYTHON_PROC*)&py3_PyObject_HasAttrString},
     {"PyObject_SetAttrString", (PYTHON_PROC*)&py3_PyObject_SetAttrString},
```

**Comment by hollowmaster1496** [comment]
I've encountered the same issue. And can confirm that the 'checking python version' step of the configure script is failing.

```
checking --with-python3-command argument... /usr/bin/python3.10
checking Python version... 3.1
checking Python is 3.0 or better... yep
checking Python's abiflags...
checking Python's install prefix... /usr
checking Python's execution prefix... /usr
checking Python's configuration directory... (cached) /usr/lib/python3.10/config-3.10-x86_64-linux-gnu
...
checking Python3's dll name... libpython3.10.so.1.0
checking if -pthread should be used... yes
checking if compile and link flags for Python 3 are sane... no: PYTHON3 DISABLED
```

This issue **does not** occur with python3.9:
```
checking Python3's dll name... libpython3.9.so.1.0
checking if -pthread should be used... yes
checking if compile and link flags for Python 3 are sane... yes
checking if -fPIE can be added for Python3... yes
```

**Comment by puremourning** [comment]
I think this was fixed with:

```
commit 23c0192166760b0d73bd39252ca72e3cfe596f6e (tag: v8.2.2876)
Author: Bram Moolenaar <Bram@vim.org>
Date:   Fri May 21 11:43:58 2021 +0200

    patch 8.2.2876: configure cannot detect Python 3.10

    Problem:    Configure cannot detect Python 3.10.
    Solution:   Use sys.version_info. (closes #8233)
```
